src/python/pants/backend/experimental/visibility/register.py

```diff
@@ -0,0 +1,191 @@
+
+from pants.engine.internals.dep_rules import (
+    BuildFileDependencyRules,
+    BuildFileDependencyRulesImplementation,
+    BuildFileDependencyRulesImplementationRequest,
+    DependencyRule,
+    DependencyRuleAction,
+    DependencyRules,
+)
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import (
+    Dependencies,
+    DependenciesRuleAction,
+    DependenciesRuleActionRequest,
+    FieldSet,
+    ValidatedDependencies,
+    ValidateDependenciesRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.frozendict import FrozenDict
+
+
+class BuildFileVisibilityImplementationRequest(BuildFileDependencyRulesImplementationRequest):
+    pass
+
+
+class VisibilityValidateFieldSet(FieldSet):
+    required_fields = (Dependencies,)
+
+
+class VisibilityValidateDependenciesRequest(ValidateDependenciesRequest):
+    field_set_type = VisibilityValidateFieldSet
+
+
+@dataclass(frozen=True)
+class VisibilityRule(DependencyRule):
+    action: DependencyRuleAction
+    pattern: str
+
+    @classmethod
+    def parse(cls, rule: str) -> VisibilityRule:
+        if rule.startswith("!"):
+            action = DependencyRuleAction.DENY
+            pattern = rule[1:]
+        elif rule.startswith("?"):
+            action = DependencyRuleAction.WARN
+            pattern = rule[1:]
+        else:
+            action = DependencyRuleAction.ALLOW
+            pattern = rule
+        return cls(action, pattern)
+
+    def match(self, path: str, relpath: str) -> bool:
+        pattern = relpath if self.pattern == "." else self.pattern
+        if pattern.startswith("./"):
+            pattern = relpath + pattern[1:]
+        return fnmatch(path, pattern)
+
+
+VisibilityRules = Tuple[VisibilityRule, ...]
+
+
+@dataclass(frozen=True)
+class BuildFileVisibilityRules(BuildFileDependencyRules):
+    default: DependencyRuleAction
+    all: DependencyRules
+    targets: Mapping[str, DependencyRules]
+
+    @classmethod
+    def create(
+        cls,
+        default: DependencyRuleAction = DependencyRuleAction.ALLOW,
+        all: Iterable[str | DependencyRule] = (),
+        targets: Mapping[str, Iterable[str | DependencyRule]] = {},
+    ) -> BuildFileVisibilityRules:
+        return cls(
+            default,
+            cls.parse_visibility_rules(all),
+            FrozenDict(
+                {
+                    type_alias: cls.parse_visibility_rules(rules)
+                    for type_alias, rules in targets.items()
+                }
+            ),
+        )
+
+    @classmethod
+    def parse_visibility_rules(cls, rules: Iterable[str | DependencyRule]) -> DependencyRules:
+        return tuple(
+            VisibilityRule.parse(rule) if isinstance(rule, str) else rule for rule in rules
+        )
+
+    def get_rules(self, type_alias: str) -> VisibilityRules:
+        if type_alias in self.targets:
+            return cast(VisibilityRules, self.targets[type_alias])
+        else:
```

Comment:
So any type_alias that isn't a target type will be the same as `"all"`? Shouldn't we check for `all` and error otherwise?

Reply:
Yes, the idea being that if you provide target specific rules, use them, otherwise the `all` rules will be as default rules for any other targets. I did consider if the target specific rules should be prepended to the `all` list of rules, so that `all` always applies, unless there is a target specific rule that matches that takes precedence. I think that could make sense.

I don't follow what you suggest we should check here?

Reply:
I've tweaked this slightly for the visibility implementation so that after any target specific rules, also the rules in `all` will be considered.

Reply:
Yeah, I think `all` should always apply, even if there is also a rule for a specific target type, then both rules should apply to that target type.

A rule that applies only if there is no rule for the specific target type should be named something like `default`, but we already have a more general `default`, so I'm not sure what I would call it.  But it seems like we might need this more general default mechanism as well as "all"?

I expect people to use `all` ~95% of the time, for sanity.

Reply:
Stale thread after the re-design we implemented yesterday.